### PR TITLE
support restricting HTTP method in rules

### PR DESCRIPTION
**Pull request by eschultink**
### Features
 - support for restricting http methods by endpoint

### Change implications

 - dependencies added/changed? **no**

---
- To see the specific tasks where the Asana app for GitHub is being used, see below:
  - https://app.asana.com/0/0/1203208515850410

### CODE REVIEW

File: java/core/src/main/java/co/worklytics/psoxy/rules/Rules2.java [hunk 1]
```diff
@@ -103,6 +103,18 @@ public Optional<List<String>> getAllowedQueryParamsOptional() {
             return Optional.ofNullable(allowedQueryParams);
         }
 
+
+        //if provided, only http methods in this list will be allowed
+        @JsonInclude(JsonInclude.Include.NON_NULL)
+        List<String> allowedMethods;
```
Comment (aperez-worklytics) [inline java/core/src/main/java/co/worklytics/psoxy/rules/Rules2.java:109]:
```suggestion
        Set<String> allowedMethods;
```

File: java/core/src/main/java/co/worklytics/psoxy/rules/Rules2.java [hunk 2]
```diff
@@ -103,6 +103,18 @@ public Optional<List<String>> getAllowedQueryParamsOptional() {
...
+
+        @JsonIgnore
+        public Optional<List<String>> getAllowedMethods() {
```
Comment (aperez-worklytics) [inline java/core/src/main/java/co/worklytics/psoxy/rules/Rules2.java:112]:
```suggestion
        public Optional<Set<String>> getAllowedMethods() {
```

File: java/core/src/main/java/co/worklytics/psoxy/gateway/impl/CommonRequestHandler.java [hunk 3]
```diff
@@ -210,7 +210,7 @@ public HttpEventResponse handle(HttpEventRequest request) {
                             .withPseudonymImplementation(pseudonymImplementation.get()));
                     }
 
-                    proxyResponseContent = sanitizer.sanitize(targetUrl, responseContent);
+                    proxyResponseContent = sanitizer.sanitize("GET", targetUrl, responseContent);
```
Comment (aperez-worklytics) [inline java/core/src/main/java/co/worklytics/psoxy/gateway/impl/CommonRequestHandler.java:213]: :yellow_circle:  Use the original http method:
```suggestion
                    proxyResponseContent = sanitizer.sanitize(request.getHttpMethod(), targetUrl, responseContent);
```

All the APIs we have work with GET, but Dropbox works with POST for all methods

**Review by aperez-worklytics — APPROVED** [review]
Some comments, the important one is marked with :yellow_circle: , after that change OK for me

**Review by aperez-worklytics — CHANGES_REQUESTED** [review]
Provide the method used in the request instead of hardcoded GET

File: java/core/src/main/java/co/worklytics/psoxy/gateway/impl/CommonRequestHandler.java (hunk 3 above)
Comment (eschultink) [reply java/core/src/main/java/co/worklytics/psoxy/gateway/impl/CommonRequestHandler.java:213]: :face_palm

**Review by jlorper — APPROVED** [review]
Ok, but we could have used  the `com.google.api.client.http.HttpMethods` constants everywhere.

**Comment by eschultink** [comment]
> Ok, but we could have used the `com.google.api.client.http.HttpMethods` constants everywhere.

I want to migrate that to apache, bc Apache supports more sophisticated control of connections (https ciphers, etc)

**Pull request merged**